File: code/typescript-for-csharp-java/errors/l02_nominal.ts
// errors/l02_nominal.ts
// A #private field makes a class nominal: no other class, however similar, has that field
class Celsius {
  #value: number;
  constructor(value: number) {
    this.#value = value;
  }
  get value() {
    return this.#value;
  }
}
class Fahrenheit {
  #value: number;
  constructor(value: number) {
    this.#value = value;
  }
  get value() {
    return this.#value;
  }
}

type Kelvin = number & { readonly brand: 'Kelvin' };

const water: Celsius = new Fahrenheit(212);
const room: Kelvin = 293.15;
console.log(water.value, room);

File: code/typescript-for-csharp-java/compare_fail/L02Nominal.java
// compare_fail/L02Nominal.java
record Celsius(double value) {}
record Fahrenheit(double value) {}

public class L02Nominal {
    public static void main(String[] args) {
        Celsius water = new Fahrenheit(212);
        System.out.println(water);
    }
}
